apps/emqx/src/emqx_router_helper.erl

```diff
@@ -129,5 +123,45 @@ code_change(_OldVsn, State, _Extra) ->
 stats_fun() ->
     emqx_stats:setstat('topics.count', 'topics.max', emqx_router:stats(n_routes)).
 
+%% This is just to avoid cleaning up routes twice in case the node is down,
+%% as we can expect {mria, down}, {node, down} under normal circumstances
+%%  when a replicant node is being stopped.
+maybe_drain_nodedown_msg(What, DownNode) when DownNode =/= node(), What =/= node ->
+    receive
+        {membership, {node, down, DownNode}} -> ok
+    after ?DRAIN_TIMEOUT -> ok
+    end;
+maybe_drain_nodedown_msg(_, _) ->
+    ok.
+
 cleanup_routes(Node) ->
-    emqx_router:cleanup_routes(Node).
+    case mria_rlog:role() of
+        core ->
+            global:trans(
+                {?LOCK(Node), self()},
+                fun() -> emqx_router:cleanup_routes(Node) end,
+                %% Need to acquire the lock on replicants too,
+                %% since replicants participate in locking (see init/1).
+                [node() | nodes()],
+                %% It's important to do a limited number of retries (not infinity).
+                %% Otherwise, if a down node comes back in a short time
+                %% and wins the lock in init/1,
+                %% emqx_router_helper on another node can desperately retry it indefinitely.
+                %% Moreover, as we currently set prevent_overlapping_partitions = false,
+                %% the connection loss between two nodes may not be visible to other nodes.
+                %% For example, Node A is down for Node B but not for Node C,
+                %% Node B tries to cleanup A routes, but cannot acquire the lock,
+                %% as A's lock is still present on node C (C and A still can communicate).
+                %% If retries were set to infinity, there will be a risk for emqx_router_helper
+                %% to be blocked forever on node B, trying to acquire the lock and do cleanup.
+                2
+            );
+        replicant when Node =:= node() ->
+            %% Mnesia (or Mria) is down locally,
+            %% but emqx_router_helper is still running and holds the lock,
+            %% remove it so that other healthy nodes can do cleanup.
+            global:del_lock({?LOCK(Node), self()});
```

Comment:
Q: Hm, is it really something that can happen during normal operation? What would be the circumstances that trigger it? I'm under impression that mria / mnesia app is started in permanent mode in production releases, so it going down should cause whole node to terminate anyway.

Reply:
@zmstone what do you think about ignoring mnesia/mria down events  in the router_helper and only do cleanups when nodedown is received?